### ISSUE TITLE
Ask Casey for Feedback

### ISSUE THREAD

**Issue by gnzlbg** [gnzlbg/fixed_capacity_vector#1]
@CaseyCarter Since I don't have your email, I would like to ask you if you would be willing to read my proposal and give me some feedback. 

In particular I am not very familiar with POCMA in practice, and I think that considering it when discussing `std::vector` and `small_vector` would be helpful.

**Comment by CaseyCarter** [comment]
## Zero-sized

`sizeof(T)` reports the size of the object representation of a complete object of type `T`, and complete objects never have zero size. Consequently, nothing in this section makes sense ;)

You *could* require `is_empty<inline_vector<T, 0>>::value` to be `true`, although at this point in the paper I see no benefit to *requiring* it to be empty rather than leaving that to QoI.

I don't see how to implement `sizeof(inline_vector<T, C>) == 1` when `is_empty<T>`. Doing so would seem to require that multiple `T` objects have the same address, which the C object model forbids. 

## `constexpr` support

`std::reverse_iterator` is `constexpr` in the C++17 WP as of JAX thanks to (P0031)[http://wg21.link/p0031]. `TrivialType<T>` presumably means `is_trival<T>`?

Implementation of `inline_vector` is going to require backing storage into which elements are constructed with placement new; without core language changes, those constructions cannot be `constexpr`.

## Exception Safety

You can't provide even the basic guarantee if T's destructor can throw; I'd simply make Destructible<T> a precondition as it is for basically the entire standard library.

I strongly agree with your decision to leave the behavior undefined upon "exceeding" the capacity. `bad_alloc` doesn't make sense to me here - no allocation takes place, how could it fail? I'd leave this a precondition for now and let future language support for contracts give people the behavior they want. a brief survey of the choices made in the prior art would be useful here.

## Default Initialization

This section is unclear to me. Given that `new` expressions are forbidden in `constexpr` context, I can't see how functions that modify the size of an `inline_vector` can be `constexpr` - whether they default-initialize or not.

## Naming

No strong feelings, except a dislike for `stack_vector`: the name seems to imply a relationship with `std::stack`.

## Spec.

`size_type` is specified to be `size_t`, contradicting the statement in the Storage/Memory Layout section that "`inline_vector<T, Capacity>::size_type` is the smallest unsigned integer type that can represent `Capacity`."

`noexcept(true)` could simply be `noexcept`. LWG will almost certainly want to strip noexcept from any of the operations that have narrow contracts (The so-called "Lakos rule"). Library practice is to add `noexcept` only to operations with "wide" contracts (no preconditions). If you believe some of these noexcepts are important, be prepared to defend them.

`= deleted;` should be `= delete;`, or better yet, the deleted declarations of `reserve` and `shrink_to_fit` should be removed altogether.

Many of the complexity requirements are stated as `O(N)`, but don't specify what `N` is. O(Capacity) or O(size()) or O(last - first) would be better.

non-member `swap` is missing; the noexcept clause of member `swap` could be simplified to `is_nothrow_swappable<T>{}`.

I don't get the description of `inline_vector()`. What does "it is guaranteed that no elements will be constructed unless `value_type` models `TrivialType`, in which case this guarantee is implementation defined." mean? How could an empty `inline_vector` have constructed elements? Is the intent that `inline_vector<T, C>` where `T` is a trivial type *always* contains `C` live objects of type `T` of which only `size()` participate in the value of the object?

`explicit inline_vector(size_type n)` should specify "default-initialized" or "value-initialized" elements instead of "default constructed." "throws `bad_alloc` if `\p n > capacity()`" seems to contradict the earlier statement that exceeding capacity is a precondition violation.

For `inline_vector(InputIterator first, InputIterator last)`, the requirement you want is "`value_type` shall be `EmplaceConstructible` into `*this` from `*first`"; the current formulation misses the case that the iterator's reference type is not a reference type. You need also require that InputIterator meets the requirements for an input iterator.

You could replace the entire description of `inline_vector(initializer_list<value_type> il)` with "Effects: Equivalent to `inline_vector(il.begin(), il.end())`"

## Assignment
There's some future work leaks here: assignment operators from `inline_vector`s with differing capacities.

## Destruction
I don't think you need this section at all if you omit/declare the destructor as defaulted in the class synposis.

## Iterators
Here (and elsewhere) I would simply omit "Requirements: none" and "Enabled: always." I admit this is stylistic.

Again, the {const_}reverse_iterator returning functions can be constexpr.

The vague statement on the relationship of `data` with the iteration functions should be either ommitted or expanded into a note with a more complete explanation.

## Size/capacity
This section doesn't tell me what `size`/`capacity`/`max_size`/`empty` do; it just says "Effects: none."

## Element/data access
Again, missing operational semantics. If your intent is to pick up the meaning of some of these operations from the general sequence container requirements - and I think that's a good idea - you need to specify somewhere early in the specification that `inline_vector<T, C>` meets the sequence container requirements in [sequence.reqmts] except where otherwise specified.

`data() == begin() == end()` is only well-defined if `iterator` and `pointer` are the same type.

## Modifiers
O(# of elements inserted) is not generally implementable for e.g. `inline_vector<int, 42> v(40); v.insert(begin(), 42)`; `O(size)` elements need to be displaced.

# Summary
Mostly nitpicky things due to the specification being incomplete as I'm sure you're already aware. The big design issue that I think needs to be addressed is the constexpr/TrivialType thing. If the intent is that an `inline_vector<T, C>` of trivial type `T` always carries a fully constructed array of `C` elements, that isn't quite coming across. Also some of the `O(size)` complexity requirements would consequently need to be `O(capacity)`, e.g., for copy construction.

**Comment by CaseyCarter** [comment]
> In particular I am not very familiar with POCMA in practice, and I think that considering it when discussing std::vector and small_vector would be helpful.

I don't see the connection with POCMA: this proposal has nothing to do with allocators.

**Comment by gnzlbg** [comment]
Hey @CaseyCarter , thank you very much for the feedback! It helps a lot! 

I haven't have time to review all of it but I've went through some of the points (and filled issues for the ones I haven't worked on yet so that I move to them later).

> sizeof(T) reports the size of the object representation of a complete object of type T, and complete objects never have zero size. Consequently, nothing in this section makes sense ;)
>
>You could require is_empty<inline_vector<T, 0>>::value to be true, although at this point in the paper I see no benefit to requiring it to be empty rather than leaving that to QoI.

Indeed. Fixed this to require `is_empty<inline_vector<T, 0>>::value == true`. I need to specify the behavior of `begin/end/data` for vectors with zero capacity so I might as well just require it (since everyone should be implementing it this way). I can relax it later to QoI if people are really against this. (I still need to define `being == end == data` differently due to them having possibly different types).

> I don't see how to implement sizeof(inline_vector<T, C>) == 1 when is_empty<T>. Doing so would seem to require that multiple T objects have the same address, which the C object model forbids.

Yeah this is did not make any sense. I was thinking that in this case `inline_vector` becomes `view::repeat_n(T{}, size())`, so there wouldn't be any object stored (which one cannot modify anyways). However rereading how I defined data/begin/end in this case, this doesn't make any sense, and it is not worth the trouble, so I killed it. 

> Implementation of inline_vector is going to require backing storage into which elements are constructed with placement new; without core language changes, those constructions cannot be constexpr.

For `is_trivial<T>` the implementation uses a C array: `T [Capacity]` to store the values and has an implicit destructor, so it skips the usage of placement new. I am relying on this as well for making functions like `resize` `constexpr`. 

Am I allowed to do this? The elements are always constructed and never destroyed, but is this "observable" by an user? The types have trivial constructors and destructors, so trying to "observe" this somehow (e.g. by providing user-defined constructors/destructors) should result in `is_trivial<T>` being false.

> or better yet, the deleted declarations of reserve and shrink_to_fit should be removed altogether.

Those replacing `std::vector` with `inline_vector` might inadvertently try to call these functions. I thought that the error message from a deleted function is clearer (and something like `=delete("reason")` would be even better). I have no strong feelings about this though.

> I don't get the description of inline_vector(). What does "it is guaranteed that no elements will be constructed unless value_type models TrivialType, in which case this guarantee is implementation defined." mean? How could an empty inline_vector have constructed elements? Is the intent that inline_vector<T, C> where T is a trivial type always contains C live objects of type T of which only size() participate in the value of the object?

Yes, that is currently how `inline_vector` can be made `constexpr` when T is a trivial type. Whether this is a good idea or not I am still not sure. Allowing placement new and explicit destructor calls in `constexpr` is the "right" way to do this, but I tried to do what I could with the current language.

> explicit inline_vector(size_type n) should specify "default-initialized" or "value-initialized" elements instead of "default constructed." "throws bad_alloc if \p n > capacity()" seems to contradict the earlier statement that exceeding capacity is a precondition violation.

Indeed, value-initialized seems to be the right thing to do here to remain consistent with `std::vector`. Fixed.

> You could replace the entire description of inline_vector(initializer_list<value_type> il) with "Effects: Equivalent to inline_vector(il.begin(), il.end())"

Indeed, thanks!

> The big design issue that I think needs to be addressed is the constexpr/TrivialType thing. If the intent is that an inline_vector<T, C> of trivial type T always carries a fully constructed array of C elements, that isn't quite coming across.

I think that I will address this explicitly in the constexpr section (and explain how it is implemented). I don't know if it is worth it implementing this right now. It is a nice thing for a lot of types but it might change the semantics enough to result in subtle pitfalls (although I cannot think of any, i am sure others will). The "right solution" is to make `std::aligned_storage` usable in constant expressions. For that to happen at least three things that are currently disallowed in constant expressions will need to be allowed: placement new, reinterpret cast, and explicit destructor calls. This would probably allow implementing variant/optional with aligned storage instead of with "clever" recursive unions.

> I don't see the connection with POCMA: this proposal has nothing to do with allocators.

My main question about POCMA is "how does the exception safety guarantees of standard vector operations change when the memory allocated cannot be moved? (e.g. because one uses a stack allocator?). 

`std::vector` then needs to allocate new memory (which can throw), move all elements from the old to the new memory (where each move can throw), and free the old memory (which cannot throw). What happens when moving one element throws? At most the basic guarantee can be provided (trying to move elements back might result in a double throw so that's a no-no).

These are the same semantics that `inline_vector` provides, so I thought I should mention this in the comparison with `std::vector`. 

Finally, thanks thanks thanks a lot for the feedback! I still need to do a lot of work here but once I finish addressing the issues I'll put you in the acknowledgements!

**Comment by CaseyCarter** [comment]
>> I don't see how to implement `sizeof(inline_vector<T, C>) == 1` when `is_empty<T>`. Doing so would seem to require that multiple `T` objects have the same address, which the C++ object model forbids.

> Yeah this is did not make any sense. I was thinking that in this case `inline_vector` becomes `view::repeat_n(T{}, size())`, so there wouldn't be any object stored (which one cannot modify anyways). However rereading how I defined data/begin/end in this case, this doesn't make any sense, and it is not worth the trouble, so I killed it.

There may be a container design for which that makes sense, but anything with `vector` in the name basically *must* be contiguous, which would not be the case for that design. Speaking of which, I don't recall wording in the paper that asserts that specializations of `inline_vector` are *contiguous containers* per N4594 [container.requirements.general]/13; I assume that is intended.

>> Implementation of `inline_vector` is going to require backing storage into which elements are constructed with placement new; without core language changes, those constructions cannot be `constexpr`.

> For `is_trivial<T>` the implementation uses a C array: `T[Capacity]` to store the values and has an implicit destructor, so it skips the usage of placement new. I am relying on this as well for making functions like `resize` `constexpr`.

> Am I allowed to do this? The elements are always constructed and never destroyed, but is this "observable" by an user? The types have trivial constructors and destructors, so trying to "observe" this somehow (e.g. by providing user-defined constructors/destructors) should result in `is_trivial<T>` being false.

As I read through the paper, I was thinking this didn't agree with the complexity clauses that require, e.g., "exactly size() calls to `T`'s copy constructor." But yes, I agree: If `T` is trivial, the number of calls to a onstructor or destructor is not observable. 

>> I don't get the description of inline_vector(). What does "it is guaranteed that no elements will be constructed unless value_type models TrivialType, in which case this guarantee is implementation defined." mean? How could an empty inline_vector have constructed elements? Is the intent that inline_vector where T is a trivial type always contains C live objects of type T of which only size() participate in the value of the object?

> Yes, that is currently how `inline_vector` can be made `constexpr` when `T` is a trivial type. Whether this is a good idea or not I am still not sure. Allowing placement new and explicit destructor calls in constexpr is the "right" way to do this, but I tried to do what I could with the current language.

I would remove the "it is guaranteed...implementation defined" text and simply assert the postcondition that `size() == 0`. 

>> The big design issue that I think needs to be addressed is the constexpr/TrivialType thing. If the intent is that an inline_vector of trivial type T always carries a fully constructed array of C elements, that isn't quite coming across.

> I think that I will address this explicitly in the constexpr section (and explain how it is implemented). I don't know if it is worth it implementing this right now. It is a nice thing for a lot of types but it might change the semantics enough to result in subtle pitfalls (although I cannot think of any, i am sure others will). The "right solution" is to make std::aligned_storage usable in constant expressions. For that to happen at least three things that are currently disallowed in constant expressions will need to be allowed: placement new, reinterpret cast, and explicit destructor calls. This would probably allow implementing variant/optional with aligned storage instead of with "clever" recursive unions.

I would leave it in for now: arrays of trivial types are extremely common, I think the benefit of the special case justifies the additional specification load.

>>I don't see the connection with POCMA: this proposal has nothing to do with allocators.

> My main question about POCMA is "how does the exception safety guarantees of standard vector operations change when the memory allocated cannot be moved? (e.g. because one uses a stack allocator?).

> std::vector then needs to allocate new memory (which can throw), move all elements from the old to the new memory (where each move can throw), and free the old memory (which cannot throw). What happens when moving one element throws? At most the basic guarantee can be provided (trying to move elements back might result in a double throw so that's a no-no).

> These are the same semantics that inline_vector provides, so I thought I should mention this in the comparison with std::vector.

Now I see: the move behavior of `inline_vector` is element-wise. I think I'd avoid any discussion of allocators - which few people understand anyway - and focus on which of the semantics are "more like std::vector" and which are "more like std::array".

**Comment by gnzlbg** [comment]
> There may be a container design for which that makes sense, but anything with vector in the name basically must be contiguous, which would not be the case for that design. 

Yes I agree here.

> Speaking of which, I don't recall wording in the paper that asserts that specializations of inline_vector are contiguous containers per N4594 [container.requirements.general]/13; I assume that is intended.

Yes it is intended. I've filled an issue for that (ContiguousContainer and hence ContiguousIterator), need to mention this.

> I would remove the "it is guaranteed...implementation defined" text and simply assert the postcondition that size() == 0.

I think that adding the pre/post conditions is a great idea, I need to do this anyways for the wide/narrow contracts + noexcept issue. 

There is some discussion in SG14 about making it crystal clear when allocations do happen in containers and when a container is guaranteed never to allocate. While `inline_vector` never allocates, it might be that `T` does allocate on construction, so I wanted to guarantee the exact number of constructor calls that must happen. I don't know if the post-condition that `size() == 0` is enough for this, so I went with `N` constructor calls. At the same time, `N` constructor calls prevents the `is_trivial<T>` optimization for which `Capacity` constructor calls must happen.

I'd think I will go with the "larger" specification here: `N` constructor calls for `!is_trivial<T>` and `Capacity` constructor calls for `is_trivial<T>` + post: `size() == N`. That might help better understanding how the "optimization" for `is_trivial<T>` works, and it can always be left "unspecified" later.

>  the move behavior of inline_vector is element-wise. I think I'd avoid any discussion of allocators - which few people understand anyway - and focus on which of the semantics are "more like std::vector" and which are "more like std::array".

I'll improve the comparison with this. Describing the move behavior as element-wise (like `std::array`) is a very clear way of explaining the semantics, thanks! 

And really thank you a lot for your time Casey!
